Implements delivery time and order confirmation feature and its tests

amandaarruda opened the pull request.

amandaarruda committed:
feat: creation of the notifications feature

amandaarruda committed:
feat: creation of the shipping feature

amandaarruda committed:
Shipping and Notifications features updates

amandaarruda committed:
New test scenario

amandaarruda committed:
Merge branch 'BrenoCinUfpe:main' into dev

amandaarruda committed:
Delete features/notifications.feature

amandaarruda committed:
Delete features/shipping.feature

amandaarruda committed:
feat(order confirmation): creation of the feature

amandaarruda committed:
feat(delivery time): creation od the feature

amandaarruda committed:
feat(order confirmation): implement orders feature

creates the base structure of the orders feature and implements the order confirmation by email feature and its tests

amandaarruda committed:
feat(delivery time): implements delivery feature

creates and implements the feature to estimate delivery time and its tests

BrenoRev commented on src/modules/delivery/delivery.controller.ts:
```diff
@@ -0,0 +1,21 @@
+        await this.correiosService.calculateDeliveryTime(cepDestino);
+      return { tempoEntrega };
+    } catch (error) {
+      throw new BadRequestException(error.message);
```
Nem sempre todas as requisições falham por conta de requisições mal formadas.
Não precisa ser try catch a nível de controller, isso é tratado a nível de serviço e retornado o erro com status adequado.

BrenoRev commented on src/modules/delivery/delivery.module.ts:
```diff
@@ -0,0 +1,23 @@
+  imports: [
+    SoapModule.register({
+      clientName: 'SOAP_CORREIOS',
+      uri: 'http://ws.correios.com.br/calculador/CalcPrecoPrazo.asmx?wsdl',
```
Criar no .env a url do serviço e utilizar o process.env.API_CORREIOS_URL por exemplo e colocar esse valor no .env-example também para indicar para os próximos que forem utilizar qual valor devem utilizar nessa env.

BrenoRev commented on src/modules/delivery/delivery.service.spec.ts:
```diff
@@ -0,0 +1,50 @@
+    httpService = module.get<HttpService>(HttpService);
+  });
+
+  it('should be defined', () => {
```
Teste desnecessário

BrenoRev commented on src/modules/delivery/delivery.service.spec.ts:
```diff
@@ -0,0 +1,50 @@
+
+  it('should calculate delivery time', async () => {
+    const cepDestino = '50030-230';
+    const tempoEntrega = await service.calculateDeliveryTime(cepDestino);
```
Muito bom, há outros tipos de validações que podem ser feitas, por exemplo:
Validar se a função que foi mockada foi chamada com os parametros X e Y

BrenoRev commented on src/modules/delivery/delivery.service.spec.ts:
```diff
@@ -0,0 +1,50 @@
+  it('should throw BadRequestException on error', async () => {
+    const cepDestino = '00000-000';
+    jest.spyOn(httpService, 'get').mockImplementationOnce(() => {
+      throw new BadRequestException('Erro ao calcular prazo de entrega');
```
Bom teste, mas aqui há um problema:
Imagina que a API dos correios retornou erro 400 por algum mótivo qualquer, vamos imaginar que foi porque o serviço deles caiu...
e dai você retorna pro usuário que deu erro 400, está certo?
Imagina você como usuário consumindo uma API RestFul fazendo uma requisição e recebendo status 400...você vai imaginar que é porque a sua requisição está errada e não porque um serviço interno falhou.

O melhor cenário seria: Você retornar um erro 500 de "Internal server error" para indicar que ocorreu algo no processamento, mas foi por conta do servidor e dos seus serviços e não um 400 de "Requisição está errada".

BrenoRev commented on src/modules/delivery/delivery.service.ts:
```diff
@@ -0,0 +1,32 @@
+        .get(`https://api.correios.com.br/${cep}`)
+        .toPromise();
+
+      const prazoEntrega = response.data.prazoEntrega;
```
Utilizar optional chaining
exemplo: response?.data?.prazoEntrega;

Why?
Porque imagina que a response.data veio vázio e você tentou acessar a terceira propriedade, vai entrar no cenário de "null.prazoEntrega" e o servidor vai lançar um internal server error antes de chegar na validação do if(!prazoEntrega)

BrenoRev commented on src/modules/delivery/delivery.service.ts:
```diff
@@ -0,0 +1,32 @@
+      const prazoEntrega = response.data.prazoEntrega;
+
+      if (!prazoEntrega) {
+        throw new BadRequestException(
```
Pode lançar um InternalServerErrorException
Um badrequest indica que o usuário fez uma requisição errada, e nesse caso o que aconteceu de erro foi do servidor externo.

amandaarruda committed:
feat(email): update email.service.ts

BrenoRev commented on src/modules/delivery/delivery.service.ts:
```diff
@@ -0,0 +1,32 @@
+
+      return parseInt(prazoEntrega);
+    } catch (error) {
+      if (error.response && error.response.status === 404) {
```
Aqui você teria que utilizar a função que criei "handleError"
Porque dentro dele tem a seguinte lógica:
Se for uma exceção que você mandou em código como você fez no throw new BadRequestException
Ele vai relançar pro cliente, mas se for uma exceção não tratada vai lançar um internal server error com uma mensagem padrão

BrenoRev commented on src/modules/delivery/steps/delivery.steps.ts:
```diff
@@ -0,0 +1,104 @@
+        data: { prazoEntrega: 5 },
+      };
+      mockHttpService.get.mockReturnValue(of(mockResponse));
+      try {
```
Remover try catch

BrenoRev commented on src/modules/delivery/steps/delivery.steps.ts:
```diff
@@ -0,0 +1,104 @@
+          },
+        })),
+      );
+      try {
```
remover try catch, senão você nunca vai saber se é um falso positivo ou um teste realmente falhando

BrenoRev commented on src/modules/orders/orders.controller.spec.ts:
```diff
@@ -0,0 +1,19 @@
+    controller = module.get<OrdersController>(OrdersController);
+  });
+
+  it('should be defined', () => {
```
Teste desnecessário

amandaarruda committed:
feat: add database connection and disconnection on module init and destroy

BrenoRev commented on src/modules/orders/orders.service.spec.ts:
```diff
@@ -0,0 +1,126 @@
+  });
+
+  describe('createOrder', () => {
+    it('deve criar um novo pedido e enviar email de confirmação', async () => {
```
Pedro solicitou que devemos utilizar os valores do .features para popular as variáveis

BrenoRev commented on src/modules/orders/orders.service.spec.ts:
```diff
@@ -0,0 +1,126 @@
+        },
+      });
+      expect(emailService.sendEmail).toHaveBeenCalledWith(
+        `<h1>Eba! Seu pedido foi confirmado. Agradecemos a preferência pelo seu pedido!</h1>
```
Fazer uma variável com esse valor ou pegar diretamente da classe de email

amandaarruda committed:
Add HttpModule and SoapModule and integrate Orders and Delivery modules

BrenoRev commented on src/modules/orders/orders.service.ts:
```diff
@@ -0,0 +1,36 @@
+      },
+    });
+
+    const emailMarkup = `<h1>Eba! Seu pedido foi confirmado. Agradecemos a preferência pelo seu pedido!</h1>
```
Fazer isso estar em uma classe de utils para reutilizar no teste é bacana

amandaarruda committed:
Update README.md

amandaarruda committed:
fix: add @nestjs/axios and nestjs-soap dependencies

amandaarruda committed:
fix: add @nestjs/axios and nestjs-soap dependencies

amandaarruda committed:
fix: try catch removed

amandaarruda committed:
fix: Changed the SOAP service URI to use the process.env.API_CORREIOS_URL environment variable.

amandaarruda committed:
Add API_CORREIOS_URL

amandaarruda committed:
fix: remove unnecessary 'should be defined' tests and improve error handling

Removed the it('should be defined') test, which was unnecessary.
Used optional chaining when accessing "prazoEntrega".
Used InternalServerErrorException instead of BadRequestException for invalid addresses.
Created handleError method to handle errors and throw InternalServerErrorException to indicate server problems.

amandaarruda committed:
Update delivery.service.spec.ts

Added validation to ensure the mocked function was called with the expected parameters.
Changed the thrown exception from BadRequestException to InternalServerErrorException.

amandaarruda committed:
fix: remove unnecessary 'should be defined' tests and improve error handling

Removed the it('should be defined') test, which was unnecessary.
Added validation to ensure the mocked function was called with the expected parameters.
Changed the thrown exception from BadRequestException to InternalServerErrorException.

amandaarruda committed:
fix: using optional chaining tests and improve error handling

Used optional chaining when accessing prazoEntrega.
Used InternalServerErrorException instead of BadRequestException for invalid addresses.
Created handleError method to handle errors and throw InternalServerErrorException to indicate server problems.

amandaarruda committed:
fix: try catch removed

amandaarruda committed:
fix: remove unnecessary 'should be defined' tests

amandaarruda committed:
feat(email): create email.utils.ts

amandaarruda committed:
fix: removed unnecessary tests

Removed unnecessary tests that only checked if controllers/services were defined.
Used the .feature values ​​to populate the variables in the service test.

BrenoRev reviewed (approved):
LGTM
